src/v/storage/mvlog/batch_collecting_stream_utils.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "storage/mvlog/batch_collecting_stream_utils.h"
+
+#include "base/vlog.h"
+#include "storage/mvlog/batch_collector.h"
+#include "storage/mvlog/logger.h"
+#include "storage/record_batch_utils.h"
+
+namespace storage::experimental::mvlog {
+
+std::ostream& operator<<(std::ostream& o, collect_stream_outcome out) {
+    switch (out) {
+    case collect_stream_outcome::buffer_full:
+        return o << "collect_stream_outcome::buffer_full";
+    case collect_stream_outcome::end_of_stream:
+        return o << "collect_stream_outcome::end_of_stream";
+    case collect_stream_outcome::stop:
+        return o << "collect_stream_outcome::stop";
+    }
+}
+
+ss::future<result<collect_stream_outcome, errc>>
+collect_batches_from_stream(entry_stream* entries, batch_collector* collector) {
+    while (true) {
+        auto entry_res = co_await entries->next();
+        if (entry_res.has_error()) {
+            co_return entry_res.error();
+        }
+        auto& entry = entry_res.value();
+        if (!entry.has_value()) {
+            co_return collect_stream_outcome::end_of_stream;
+        }
+        switch (entry->hdr.type) {
+        case entry_type::record_batch: {
+            auto res = collect_batch_from_buf(
+              std::move(entry->body), collector);
+            if (res.has_error()) {
+                co_return res.error();
+            }
+            switch (res.value()) {
+            // Signals that end the stream should be passed through to the
+            // caller so the stream can be reset...
```

Comment:
`so the stream can be reset` confused about the meaning of this clause. buffer_full sounds like it is time to drain the collector and continue reading...

Reply:
Ah true, I think I wrote this comment when the outcomes looked a bit different. Will fix